File: oauth2.py
import requests
import os
from urllib.parse import quote

API_ENDPOINT = 'https://discord.com/api/v10'

def exchange_code(code: str, redirect_uri: str, client_id: str, client_secret: str):
  """
  https://discord.com/developers/docs/topics/oauth2#authorization-code-grant-access-token-exchange-example

  """
  data = {
    'client_id': client_id,
    'client_secret': client_secret,
    'grant_type': 'authorization_code',
    'code': code,
    'redirect_uri': redirect_uri
  }
  headers = {
    'Content-Type': 'application/x-www-form-urlencoded'
  }
  r = requests.post('%s/oauth2/token' % API_ENDPOINT, data=data, headers=headers)
  r.raise_for_status()
  return r.json()

def make_authorization_url(client_id: str, redirect_uri: str) -> str:
  return f"https://discord.com/oauth2/authorize?response_type=code&client_id={client_id}&scope=identify%20guilds.join%20guilds.members.read&redirect_uri={quote(redirect_uri)}"

File: oauth2.ts
import { createJWT, getCookies, verifyJWT } from "./deps.ts";

/**
 * handle handles request to /oauth endpoint.
 */
export async function handle(
  request: Request,
  clientID: string,
  clientSecret: string,
  jwtSecret: string,
  redirectURI: string,
): Promise<Response> {
  const url = new URL(request.url);
  const existingJWT = getCookies(request.headers)["jwt"];
  if (existingJWT) {
    const accessToken = await verifyJWT(existingJWT, jwtSecret);
    if (accessToken) {
      return new Response(null, {
        status: 302,
        headers: { Location: "/forms" },
      });
    }
  }

  // Handle unauthorized requests.
  const code = url.searchParams.get("code");
  if (!code) {
    return Response.redirect(makeAuthorizationURL(clientID, redirectURI));
  }

  // Exchange code for access token.
  const accessToken = await exchangeCode(
    code,
    clientID,
    clientSecret,
    redirectURI,
  );

  // Encode access token into JWT.
  const jwt = await createJWT(
    { alg: "HS512", typ: "JWT" },
    { accessToken },
    // TODO: Figure out how to convert a string into
    // a CryptoKey before passing to this function.
    jwtSecret,
  );

  // Set JWT as a cookie and redirect to /forms.
  return new Response(null, {
    status: 302,
    headers: {
      "Set-Cookie": `jwt=${jwt}; Path=/; HttpOnly`,
      Location: "/forms",
    },
  });
}

/**
 * makeAuthorizationURL creates a Discord authorization URL.
 */
export function makeAuthorizationURL(
  clientID: string,
  redirectURI: string,
): string {
  return `https://discord.com/oauth2/authorize?response_type=code&client_id=${clientID}&scope=identify%20guilds.members.read&redirect_uri=${redirectURI}`;
}

/**
 * exchangeCode exchanges a code for a Discord access token.
 */
export async function exchangeCode(
  code: string,
  clientID: string,
  clientSecret: string,
  redirectURI: string,
): Promise<string> {
  const r = await fetch(`${API_ENDPOINT}/oauth2/token`, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      client_id: clientID,
      client_secret: clientSecret,
      grant_type: "authorization_code",
      code,
      redirect_uri: redirectURI,
    }),
  });
  const json = await r.json();
  return json.access_token;
}

const API_ENDPOINT = "https://discord.com/api/v10";
